### PR TITLE
Add the number of components

### PR DESCRIPTION
**Pull request by jajoe**
Add a parameter number of components to the function predict to enable dimensionality reduction

### CODE REVIEW

File: src/pca.js [hunk 1]
```diff
@@ -7,6 +7,14 @@ const Stat = require('ml-stat').matrix;
 const mean = Stat.mean;
 const stdev = Stat.standardDeviation;
 
+function xrange(n){
```
Comment (maasencioh) [inline src/pca.js]: use https://www.npmjs.com/package/lodash.range instead

File: src/pca.js [hunk 2]
```diff
@@ -87,22 +95,32 @@ class PCA {
...
 
+
     /**
      * Project the dataset into the PCA space
      * @param {Matrix} dataset
+     * @param {number} number of components
      * @return {Matrix} dataset projected in the PCA space
      */
-    predict(dataset) {
+    predict(dataset, numberComponents = -1) {
```
Comment (maasencioh) [inline src/pca.js]: is better to use a options object

**Comment by jajoe** [comment]
Except a problem with Node 4, everything seems to be OK. May I merge @targos ?

File: src/pca.js [hunk 3]
```diff
@@ -87,22 +87,36 @@ class PCA {
         return new PCA(true, model);
     }
 
+
     /**
      * Project the dataset into the PCA space
      * @param {Matrix} dataset
+     * @param {number} number of components
```
Comment (targos) [inline src/pca.js]: this needs to be updated

File: src/pca.js [hunk 4]
```diff
@@ -87,22 +87,36 @@ class PCA {
...
      * @return {Matrix} dataset projected in the PCA space
      */
-    predict(dataset) {
-        dataset = new Matrix(dataset);
+    predict(dataset, options = {}) {
+        const {
+           nComponents = -1
```
Comment (targos) [inline src/pca.js]: this can default to `this.U.columns` instead of a magic value

File: src/pca.js [hunk 5]
```diff
@@ -87,22 +87,36 @@ class PCA {
...
                 dataset.divRowVector(this.stdevs);
             }
         }
 
-        return dataset.mmul(this.U);
+        var predictions = dataset.mmul(this.U);
+        if (nComponents === -1) {
+            return predictions;
+        } else {
+            if (nComponents <= predictions.columns) {
```
Comment (targos) [inline src/pca.js]: I'd rather check this before doing the multiplication and throw if the parameter is out of range

**Comment by targos** [comment]
If you rebase your branch on master, it will fix the test issue.

**Comment by jajoe** [comment]
I haven't done the subMatrix before the mmul, but I find the source code seems to be clean (readable). Tell me if you're OK for a merge @targos.

**Comment by targos** [comment]
thank you

**Comment by targos** [comment]
Published in v2.1.0

**Pull request merged**